README.md
# jupiter-node-sdk

A thin wrapper over the [Jupiter blockchain API](https://jpr.gojupiter.tech/test) with some utility functions to make doing common tasks like fetching data and encrypting/decrypting data easy.

## Install

`$ npm install -s jupiter-node-sdk`

## Usage

```ts
import JupiterClient from 'jupiter-node-sdk'

const client = JupiterClient({
  server: `https://jpr.gojupiter.tech`,
  address: `JUP-XXX-XXX`,
  passphrase: `your secret phrase here`,
  encryptSecret: `optional alphanumeric password to encrypt data going to blockchain`, // DEFAULT: passphrase
})

const jupAmount = client.nqtToJup(stringOfNqtAmount)

const nqtAmount = client.jupToNqt(stringOfJupAmount)

const cipherText = await client.encrypt(
  `plain text to encrypt with AES-256 bit encryption`
)
const plainDecryptedText = await client.decrypt(`cipher text`)

// secretPhrase is basically ERC-20 mnemonic that you can build yourself
const {
  address,
  publicKey,
  requestProcessingTime,
  account,
} = await client.createNewAddress(secretPhrase)

const balanceJup = await client.getBalance(`JUP-XXX-XXX`).balanceNQT

// sender is address in constructor above
await client.sendMoney(recipientJupAddress, amountJupToSend)

// store any data you want (object should be a Javascript object) on the blockchain
await client.storeRecord(object)

// gets all confirmed AND unconfirmed transactions
const txns = await client.getAllTransactions()

// to make any generic Jupiter/NXT API request you can use #request as shown below
// which uses axios under the hood
await client.request('get', '/nxt', {
  params: {
    requestType: 'getBlockchainTransactions',
    account: `JUP-XXX-XXX`,
    withMessage: true,
    type: 1,
  },
})
```


Changelog.md
# Changelog
All notable changes to this project will be documented in this file.

## [Unreleased]

## [0.4.1] - 2020-08-09
### Changed
- Changed getBalance response returning confirmed and unconfirmed balance
- Added calculateExpectedFees function

## [0.4.0] - 2020-07-25
### Changed
- Increased fee for jupiter-fs binary transactions

### Added
- Added function to get all the transactions with jupiter-fs metadata

## [0.3.1] - 2020-07-15
### Changed
- Update getAllTransactions function to support jupiter metis subtypes

## [0.3.0] - 2020-07-15
### Changed
- Update storeRecord function to support jupiter metis subtypes

## [0.2.1] - 2020-07-07
### Added
- Added Changelog file
- Added getTransaction request

### Changed
- Update metadata after fork from [moontography github](https://github.com/moontography/jupiter-node-sdk)
- Transaction fee now is calcualated dynamically based on the ancrypted message size

## [0.2.0] - 2020-07-04
### Added

### Changed
- Update encryption algo to aes-256-cbc

### Removed


src/JupiterClient.spec.ts
import assert from 'assert'
import JupiterClient from './JupiterClient'

describe('JupiterClient', function() {
  this.timeout(20000)

  const myAddy = 'JUP-TUWZ-4B8Z-9REP-2YVH5'
  const client = JupiterClient({
    server: 'http://104.131.166.136:6876/test',
    address: myAddy,
    passphrase: '',
    encryptSecret: '',
  })

  describe('#createNewAddress()', function() {
    it(`should create a new address`, async () => {
      const { address, publicKey } = await client.createNewAddress('the seed')
      assert.strictEqual(typeof address, 'string')
      assert.strictEqual(typeof publicKey, 'string')
    })
  })

  describe('#getAllTransactions()', function() {
    it(`should get all transactions of type 0`, async () => {
      const txns = await client.getAllTransactions(false, 0)
      assert.strictEqual(txns.length > 0, true)
      assert.strictEqual(txns.reverse()[0].recipientRS, myAddy)
    })
  })
})


src/Encryption.ts
import crypto from "crypto";
import fs from "fs";
import { promisify } from "util";
import zlib from "zlib";

const readFile = fs.promises.readFile;
const inflate: any = promisify(zlib.inflate);
const deflate: any = promisify(zlib.deflate);

interface IEncryptionOptions {
  secret: string;
}

export default function Encryption(options: IEncryptionOptions) {
  const alg = "aes-256-cbc";
  const sec = options.secret;

  return {
    _algorithm: alg,
    _secret: sec,

    async encrypt(input: Buffer | string) {
      const secret = getFilledSecret(this._secret);
      const { iv, key } = getKeyAndIV(secret);
      const cipher = crypto.createCipheriv(this._algorithm, key, iv);

      const inputStr =
        input instanceof Buffer ? input.toString("base64") : input;
      let cipherText = cipher.update(inputStr, "utf8", "base64");
      cipherText += cipher.final("base64");
      return await this.parseData(`${cipherText}:${iv.toString("base64")}`);
    },

    async encryptFileUtf8(filePath: string) {
      const fileText = await readFile(filePath, { encoding: "utf8" });
      return await this.encrypt(fileText);
    },

    async decrypt(text: string) {
      const inflatedString = (await this.parseData(text, false)).toString();
      const [rawBase64, ivBase64] = inflatedString.split(":");
      const iv = Buffer.from(ivBase64, "base64");
      const secret = getFilledSecret(this._secret);
      const { key } = getKeyAndIV(secret, iv);
      const decipher = crypto.createDecipheriv(this._algorithm, key, iv);

      let dec = decipher.update(rawBase64, "base64", "utf8");
      dec += decipher.final("utf8");
      return dec;
    },

    async decryptFileUtf8(filePath: string) {
      const fileText = await readFile(filePath, { encoding: "utf8" });
      return await this.decrypt(fileText);
    },

    async fileToHash(filePath: string) {
      return await new Promise((resolve, reject) => {
        const sha256Sum = crypto.createHash("sha256");

        const s = fs.createReadStream(filePath);
        s.on("data", (data) => sha256Sum.update(data));
        s.on("error", reject);
        s.on("end", () => resolve(sha256Sum.digest("base64")));
      });
    },

    // Handles any gzip/deflating/inflating we might be doing to data
    // we're passing to and from Redis.
    // NOTE: if inflating, we will always return a raw Buffer. If deflating,
    // we return a base64 encoded string.
    async parseData(
      value: string,
      isRawData: boolean = true
    ): Promise<Buffer | string> {
      switch (isRawData) {
        case false:
          return await inflate(Buffer.from(value, "base64"));

        default:
          //true
          const compressedValue = await deflate(value);
          return Buffer.from(compressedValue).toString("base64");
      }
    },
  };
}

// Private methods
function getFilledSecret(secret: string): string {
  const sha256Sum = crypto.createHash("sha256");
  sha256Sum.update(secret);
  return sha256Sum.digest("base64");
}

function getKeyAndIV(key: string, iv?: Buffer) {
  const ivBuffer = iv || crypto.randomBytes(16);
  const derivedKey = crypto.pbkdf2Sync(key, ivBuffer, 1e5, 32, "sha256");
  return {
    iv: ivBuffer,
    key: derivedKey,
  };
}


src/utils/index.ts
import WordList from './WordList'

export async function sleep(milliseconds = 1000) {
  return await new Promise((resolve) => setTimeout(resolve, milliseconds))
}

// The following is the code which will generate a list of 12
// random words that will be used to generate an nxt account
export function generatePassphrase(): string {
  return getWords(12).join(' ')
}

// The following is the code which will generate a list of 4
// random words that will be used to generate an api key
export function generateKeywords(): string {
  return getWords(4).join(' ')
}

function getWords(numWords: number): string[] {
  const list = WordList()
  const seedphrase: string[] = new Array(numWords).fill('').reduce((words) => {
    const word = list.splice(Math.floor(Math.random() * list.length), 1)
    return words.concat([word])
  }, [])
  return seedphrase
}


src/JupiterClient.ts
import assert from 'assert'
import axios, { AxiosResponse } from 'axios'
import qs from 'querystring'
import BigNumber from 'bignumber.js'
import Encryption from './Encryption'
import { sleep } from './utils'

export default function JupiterClient(opts: IJupiterClientOpts) {
  const encryption = Encryption({
    secret: opts.encryptSecret || opts.passphrase,
  })
  const CONF = {
    feeNQT: opts.feeNQT || 150,
    deadline: opts.deadline || 60,
    minimumFndrAccountBalance: opts.minimumFndrAccountBalance || 50000,
    minimumUserAccountBalance: opts.minimumUserAccountBalance || 100000,
    jupNqtDecimals: opts.jupNqtDecimals || 8,
  }

  const TYPE_DATA_FS = 8;
  const SUBTYPE_MESSAGING_METIS_METADATA_TYPE = 0;

  function setClient(serverUrl: string = opts.server) {
    return axios.create({
      baseURL: serverUrl,
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
        'User-Agent': 'jupiter-node-sdk',
      },
    })
  }

  return {
    recordKey: opts.recordKey,

    // export all options data provided in object returned
    ...opts,

    config: CONF,

    client: setClient(opts.server),
    lastClientTest: 0,
    peerList: [{ announcedAddress: opts.server }],

    // balances from the API come back as NQT, which is 1e-8 JUP
    nqtToJup(nqt: string): string {
      return new BigNumber(nqt).div(CONF.jupNqtDecimals).toString()
    },

    jupToNqt(jup: string): string {
      return new BigNumber(jup).times(CONF.jupNqtDecimals).toString()
    },

    calculateMessageFee(encryptedMessageLength: number): number {
      let fee = CONF.feeNQT
      if (!encryptedMessageLength)
          return fee
      if (encryptedMessageLength <= 5000) {
        fee = 16500
      } else if (encryptedMessageLength <= 10000) {
        fee = 32000
      } else if (encryptedMessageLength <= 20000) {
        fee = 63000
      } else if (encryptedMessageLength <= 30000) {
        fee = 94500
      } else if (encryptedMessageLength <= 40000) {
        fee = 126000
      } else {
        fee = 135000
      }
      return fee
    },

    calculateExpectedFees(data: Array<string>): number {
      let expectedFees = 0;
      data.forEach((data) => expectedFees += this.calculateMessageFee(data.length));
      return expectedFees;
    },

    decrypt: encryption.decrypt.bind(encryption),
    encrypt: encryption.encrypt.bind(encryption),

    setClient(...args: any[]) {
      return (this.client = setClient(...args))
    },

    async setClientRoundRobin() {
      let peerListCopy = this.peerList.slice()
      while (peerListCopy.length > 0) {
        const [peer] = peerListCopy.splice(
          Math.random() * peerListCopy.length,
          1
        )
        assert(peer, 'peer does not exist')
        const { announcedAddress } = peer
        const isConnected = await this.testConnection(announcedAddress)
        if (isConnected) {
          this.setClient(announcedAddress)
          this.peerList = await this.getPeers()
          this.lastClientTest = Date.now()
          return
        }
      }

      throw new Error(`No more peers to try and connect to`)
    },

    async testConnection(
      singleServerUrl: string,
      reasonableTimeToWaitMs: number = 5e3 // 5 second default
    ): Promise<boolean> {
      return await Promise.race([
        (async () => {
          await sleep(reasonableTimeToWaitMs)
          return false
        })(),
        (async () => {
          try {
            await this.getPeers(`${singleServerUrl}/nxt`)
            return true
          } catch (err) {
            return false
          }
        })(),
      ])
    },

    // we use this for testConnection, so allow passing in other
    // URLs for testing connections
    async getPeers(serverUrl?: string) {
      const {
        data: { peers },
      } = await this.request('get', serverUrl ? serverUrl : '/nxt', {
        params: {
          requestType: 'getPeers',
          includePeerInfo: true,
        },
        dontTest: true,
      })
      return peers
    },

    async getBalance(address: string = opts.address): Promise<object> {
      const {
        data: {
           unconfirmedBalanceNQT,
          // forgedBalanceNQT,
          balanceNQT,
          // requestProcessingTime
        },
      } = await this.request('get', '/nxt', {
        params: {
          requestType: 'getBalance',
          account: address,
        },
      })
      return {
        'balanceNQT': balanceNQT,
        'unconfirmedBalanceNQT': unconfirmedBalanceNQT
      }
    },

    async createNewAddress(passphrase: string) {
      return await this.getAddressFromPassphrase(passphrase)
    },

    async getAddressFromPassphrase(passphrase: string) {
      const {
        data: { accountRS: address, publicKey, requestProcessingTime, account },
      } = await this.request('post', '/nxt', {
        params: {
          requestType: 'getAccountId',
          secretPhrase: passphrase,
        },
      })
      return { address, publicKey, requestProcessingTime, account }
    },

    async sendMoney(recipientAddr: string, amountJup?: string) {
      const { data } = await this.request('post', '/nxt', {
        params: {
          requestType: 'sendMoney',
          secretPhrase: opts.passphrase,
          recipient: recipientAddr,
          amountNQT: amountJup
            ? amountJup
            : CONF.minimumFndrAccountBalance,
          feeNQT: CONF.feeNQT,
          deadline: CONF.deadline,
        },
      })
      if (
        data.signatureHash === null ||
        (data.errorCode && data.errorCode !== 0)
      ) {
        throw new Error(JSON.stringify(data))
      }
      return data
    },

    // async parseEncryptedRecord(cipherText: string): Promise<IFndrAccount> {
    async parseEncryptedRecord(cipherText: string): Promise<any> {
      return JSON.parse(await this.decrypt(cipherText))
    },

    // async storeRecord(record: IStringMap) {
    async storeRecord(record: any, subtype: string) {
      const finalRecordToStore = this.recordKey
        ? {
            ...record,
            [this.recordKey]: true,
          }
        : record

      const encryptedMessage = await this.encrypt(
        JSON.stringify(finalRecordToStore)
      )

      const requestType = subtype ? 'sendMetisMessage' : 'sendMessage'

      const { data } = await this.request('post', '/nxt', {
        params: {
          requestType: requestType,
          subtype: subtype,
          secretPhrase: opts.passphrase,
          recipient: opts.address,
          recipientPublicKey: opts.publicKey,
          messageToEncrypt: encryptedMessage,
          feeNQT: this.calculateMessageFee(encryptedMessage.length),
          deadline: CONF.deadline,
          compressMessageToEncrypt: true,
        },
      })
      if (data.errorCode && data.errorCode !== 0)
        throw new Error(JSON.stringify(data))
      return data
    },

    async decryptRecord(
      message: ITransactionAttachmentDecryptedMessage
    ): Promise<string> {
      const {
        data: { decryptedMessage },
      } = await this.request('get', '/nxt', {
        params: {
          requestType: 'decryptFrom',
          secretPhrase: opts.passphrase,
          account: opts.address,
          data: message.data,
          nonce: message.nonce,
        },
      })
      return decryptedMessage
    },

    async getAllMatadataTransactions(): Promise<ITransaction[]> {
      return this.getAllTransactions(true, TYPE_DATA_FS, SUBTYPE_MESSAGING_METIS_METADATA_TYPE)
    },


    async getAllTransactions(
      withMessage: boolean = true,
      type: number = 1, 
      subtype?: number
    ): Promise<ITransaction[]> {
      const [confirmed, unconfirmed] = await Promise.all([
        await this.getAllConfirmedTransactions(withMessage, type, subtype),
        await this.getAllUnconfirmedTransactions(withMessage, type),
      ])
      return unconfirmed.concat(confirmed)
    },

    async getTransaction(
      txId: string
    ): Promise<ITransaction> {
      const {
        data: {
          transaction,
        },
      } = await this.request('post', '/nxt', {
        params: {
          requestType: 'getTransaction',
          transaction: txId
        },
      })
      return transaction
    },

    async getAllConfirmedTransactions(
      withMessage: boolean = true,
      type: number = 1,
      subtype?: number
    ): Promise<ITransaction[]> {
      const {
        data: {
          transactions,
        },
      } = await this.request('get', '/nxt', {
        params: {
          requestType: 'getBlockchainTransactions',
          account: opts.address,
          withMessage,
          type,
          subtype
        },
      })
      return transactions == null ? [] : transactions
    },

    async getAllUnconfirmedTransactions(
      withMessage: boolean = true,
      type: number = 1
    ): Promise<ITransaction[]> {
      const {
        data: {
          unconfirmedTransactions,
        },
      } = await this.request('post', '/nxt', {
        params: {
          requestType: 'getUnconfirmedTransactions',
          account: opts.address,
          withMessage,
          type,
        },
      })

      return unconfirmedTransactions == null
        ? []
        : unconfirmedTransactions.reverse()
    },

    async request(
      verb: 'get' | 'post',
      path: string,
      opts?: IRequestOpts
    ): Promise<AxiosResponse> {
      // test connection every 5 minutes in case the current node dies
      if (
        (!opts || !opts.dontTest) &&
        Date.now() - this.lastClientTest >= 1000 * 60 * 5
      ) {
        assert(
          this.client.defaults.baseURL,
          `current client baseURL is not set for some reason`
        )
        const isConnected = await this.testConnection(
          this.client.defaults.baseURL
        )
        if (!isConnected) {
          await this.setClientRoundRobin()
        }
      }

      switch (verb) {
        case 'post': {
          const requestType =
            (opts && opts.params && opts.params.requestType) ||
            (opts && opts.body && opts.body.requestType)
          let body = opts && { ...opts.body, ...opts.params }
          delete body.requestType
          return await this.client.post(path, qs.stringify(body), {
            params: {
              requestType,
            },
          })
        }

        default:
          // get
          return await this.client.get(path, opts)
      }
    },
  }
}

interface IJupiterClientOpts {
  server: string
  address: string
  passphrase: string
  encryptSecret?: string
  recordKey?: string
  publicKey?: string
  feeNQT?: number
  deadline?: number
  minimumFndrAccountBalance?: number
  minimumUserAccountBalance?: number
  jupNqtDecimals?: number
}

interface IRequestOpts {
  // TODO: according to the NXT docs the only way to pass parameters is
  // via query string params, even if it's a POST. This seems bad, but for
  // now since POST body isn't support don't allow it in a request.
  params?: any
  body?: any
  dontTest?: boolean
}

interface ITransactionAttachment {
  [key: string]: any
}

interface ITransactionAttachmentDecryptedMessage {
  data: string
  nonce: string
  isText: boolean
  isCompressed: boolean
}

interface ITransaction {
  signature: string
  transactionIndex: number
  type: number
  phased: boolean
  ecBlockId: string
  signatureHash: string
  attachment: ITransactionAttachment
  senderRS: string
  subtype: number
  amountNQT: string
  recipientRS: string
  block: string
  blockTimestamp: number
  deadline: number
  timestamp: number
  height: number
  senderPublicKey: string
  feeNQT: string
  confirmations: number
  fullHash: string
  version: number
  sender: string
  recipient: string
  ecBlockHeight: number
  transaction: string
}
